Add draft SEP for Skills Extension (#69)
modelcontextprotocol/experimental-ext-skills: docs/sep-draft-skills-extension.md

    },
    {
      "name": "refunds",
      "type": "skill-md",
      "description": "Process customer refund requests per company policy",
      "url": "skill://acme/billing/refunds/SKILL.md"
    },
    {
      "type": "mcp-resource-template",
      "description": "Per-product documentation skill",
      "url": "skill://docs/{product}/SKILL.md"
    }
  ]
}
```

Index fields:

| Field | Required | Description |
|---|---|---|
| `$schema` | Yes | Schema version URI. Clients SHOULD match against known URIs before processing. |
| `skills` | Yes | Array of skill entries. |
| `skills[].type` | Yes | MUST be `"skill-md"` or `"mcp-resource-template"`. Archive distribution does not apply; supporting files are individually addressable as resources. |
| `skills[].description` | Yes | For `"skill-md"`, the skill's `description` matching its `SKILL.md` frontmatter. For `"mcp-resource-template"`, a description of the addressable skill space. |
| `skills[].url` | Yes | For `"skill-md"`, the full resource URI of the skill's `SKILL.md`. For `"mcp-resource-template"`, an [RFC 6570](https://datatracker.ietf.org/doc/html/rfc6570) URI template that resolves to `SKILL.md` resource URIs. |

> | `skills[].name` | Conditional | Required for `"skill-md"`; matches the `SKILL.md` frontmatter `name` and the final path segment. Omitted for `"mcp-resource-template"`. |

Reviewer: Is this meant to be case sensitive?  For example, I noticed that the mintlify generated skills capitalize the skill names.

Author: sorry, what are you referring to?

If you are talking about the lowercase requirement on skill name (mentioned on line 146) then this comes straight from the agent skill spec: https://agentskills.io/specification#name-field - which I do a agree is a bit weird!

Reviewer: @pja-ant yes sorry, I was running into a an issue testing an implementation of the SEP with a real skill and ran into a problem because my implementation was correctly enforcing lowercase.  Which I thought was weird.  But I will take that confusion elsewhere if its coming from the agent skills spec. :)

Clients SHOULD ignore unrecognized fields and SHOULD skip entries with an unrecognized `type`.

> **Template entries** (`type: "mcp-resource-template"`) describe a parameterized skill namespace without materializing every entry. A server SHOULD register the same `url` value as an MCP [resource template](https://modelcontextprotocol.io/specification/2025-11-25/server/resources#resource-templates) so hosts can wire template variables to the [completion API](https://modelcontextprotocol.io/specification/2025-11-25/server/utilities/completion). Hosts SHOULD surface template entries in their UI as interactive discovery points: the user fills in variables via completion, selects a skill, and the host passes the resolved URI into the conversation. This scales to servers with unbounded skill catalogs.

Reviewer 2: Wondering how would autonomous agent handle the fuzzy matching problem? It seems that clients are forced into a trial-and-error pattern or must rely on the host to provide specialized tools for discovery.

Author: It is a assumed that skill templates would be exposed via some human-in-the-loop UX (e.g. user starts typing skill name, host provides completions using MCP).

Nothing prevents a host from providing the model access to completions also, which could be interesting... but I'm not aware of any hosts that do that yet.

**Client-side** — enumerate and fetch skills:

```python
skills = await client.list_skills()               # reads skill://index.json, may be empty or absent
content = await client.read_skill_uri(
    "skill://acme/billing/refunds/SKILL.md")      # wraps resources/read, works regardless of enumeration
```

These wrappers are thin — each is a single underlying protocol call with a fixed URI pattern — but they give server authors an ergonomic way to declare skills and give client authors a discoverable entry point.

## Rationale

### Why Resources Instead of a New Primitive?

The Interest Group's [decision log](decisions.md#2026-02-26-prioritize-skills-as-resources-with-client-helper-tools) records this as settled. Skills are files; Resources exist to expose files. Reusing Resources inherits URI addressability, `resources/read`, `resources/subscribe`, templates, and the existing client tooling for free. A new primitive would duplicate most of this and add ecosystem complexity — a concern raised in [community discussion](https://github.com/modelcontextprotocol/experimental-ext-skills/issues/14).

> [SEP-2076] proposes the new-primitive alternative. That approach offers cleaner capability negotiation and dedicated list-changed notifications, but at the cost of flattening skills to name-addressed blobs — losing the directory model that the Agent Skills specification defines and that supporting files depend on.

Reviewer 2: I'm curious if using an extension here might obscure future capabilities, such as the skill-to-tool dependencies I brought up in my other comment. Would introducing a new primitive make those features easier to support?

Reviewer 3: I guess this being a convention, it keeps the door open for introducing new primitives in the future if new use cases emerge. Just my opinion

Reviewer: Just wanted to add a link here to the discussion around the other [proposal](https://github.com/modelcontextprotocol/experimental-ext-skills/pull/86) from last week: https://github.com/modelcontextprotocol/modelcontextprotocol/discussions/2628

I agree that this SEP doesn't close the door for new primitives in the future. The goal is to learn from starting with an extension and lighter weight conventions on top of resources.  One of the other interesting ideas brought up in the PR 86 discussion is using the progressive disclosure pattern from skills and applying them towards other primitives.  I think there's a lot of opportunities to experiment with that concept, whether explicitly through the protocol or by using 'skills as files.'

- **Skills are data, not directives.** Hosts MUST NOT treat skill resources as higher-authority than other context. Explicit user policy governs whether a skill is loaded at all.
- **Provenance and inspection.** Hosts SHOULD indicate which server a skill originates from when presenting it, SHOULD let users inspect a skill's content before it is loaded into model context, and MAY gate loading behind per-skill or per-server user approval.
- **Not a third-party marketplace.** This extension is for servers to ship skills that describe their own tools, not for distributing arbitrary third-party content through a connected server.

> The instructor-only scope of this extension ([decisions.md, 2026-02-14](decisions.md#2026-02-14-skills-served-over-mcp-use-the-instructor-format)) deliberately excludes the helper model. A filesystem skill might reasonably carry scripts the user has audited; an MCP skill arrives from a remote party and MUST be handled as text that influences model behavior, not as code that executes on the host.

Reviewer 2: I noticed that relying on remotely served scripts introduces a multi-hop inefficiency for the agent. MCP tools look like a great solution here. That said, the current SEP doesn't propose a dependency strategy between skills and tools (for instance, having Tool B being progressively loaded when Skill A is invoked). Are we planning to address this mapping in a future update?

Author: Hey, I think we discussed this in the previous meeting. The goal of this SEP is not to provide a progressive tool discovery mechanism for MCP and instead simply aims to be an MCP-native delivery mechanism for skills. Progressive discovery of tools is left as a harness problem to solve, which has several solutions: e.g. tool search tools, code mode, etc.

This doesn't rule out future progressive discovery mechanisms on MCP, but it is deemed out of scope for this extension.

- Every skill MUST contain a `SKILL.md` file at its root.
- `SKILL.md` MUST begin with YAML frontmatter containing at minimum the `name` and `description` fields as defined by the Agent Skills specification.
- A skill MAY contain additional files and subdirectories (references, scripts, examples, assets).

This extension does not redefine, constrain, or extend the skill format. Future revisions of the Agent Skills specification apply automatically.

### Resource Mapping

Each file within a skill directory is exposed as an MCP resource under the `skill://` scheme. The resource URI has the form:

```
skill://<skill-path>/<file-path>
```

where:

- `<skill-path>` is a `/`-separated path of one or more segments locating the skill directory within the server's skill namespace. It MAY be a single segment (`git-workflow`) or nested to arbitrary depth (`acme/billing/refunds`).
- `<file-path>` is the file's path relative to the skill directory root, using `/` as the separator.

The resource for the skill's required `SKILL.md` is therefore always addressable as `skill://<skill-path>/SKILL.md`, and the skill's root directory is the URI obtained by stripping the trailing `SKILL.md`.

> The `<skill-path>` is a locator, not an identifier. It need not match the skill's `name` (which comes from frontmatter), and servers MAY organize skills hierarchically by domain, team, version, or any other axis. Two constraints follow:

Reviewer 4: This is a bit odd that the skill name doesn't have to match the skill path, given the skill URI is used to access the resource.

Author: This is fair, originally they did match, but we wanted to support nested skills (a request from github), e.g. `skill://foo/bar/code-review/SKILL.md` and skill names cannot contain `/` so you're either forced to not have nested skills, or replace the `/` with `:`. You could perhaps use the last path segment as the skill name (`code-review`). wdyt @SamMorrowDrums

Reviewer 5: Last path segment as name seems totally reasonable to me. You cannot assume all servers are going to be able to be to find skills in certain places if you cannot provide the completions.

That would rule out skill discovery via resource templates entirely. Which is a valid approach, but it doesn't feel complex enough of an idea to reject.

Why shouldn't the spec let somebody install a skill from a known repo this way for example? GH server cannot enumerate them all.

The main alternative we discussed still has this problem, which is resource links could be returned from tools providing skill search but still need paths.

Per RFC 3986, the first segment of `<skill-path>` occupies the authority component. This carries no special semantics under this convention and clients MUST NOT attempt DNS or network resolution of it.

#### Examples

| Skill path | File | Resource URI |
|---|---|---|
| `git-workflow` | `SKILL.md` | `skill://git-workflow/SKILL.md` |
| `pdf-processing` | `references/FORMS.md` | `skill://pdf-processing/references/FORMS.md` |
| `pdf-processing` | `scripts/extract.py` | `skill://pdf-processing/scripts/extract.py` |
| `acme/billing/refunds` | `SKILL.md` | `skill://acme/billing/refunds/SKILL.md` |
| `acme/billing/refunds` | `templates/email.md` | `skill://acme/billing/refunds/templates/email.md` |

#### Resource Metadata

For each `skill://<skill-path>/SKILL.md` resource:

- `mimeType` SHOULD be `text/markdown`.
- `name` SHOULD be set from the `name` field of the `SKILL.md` YAML frontmatter.

> - `description` SHOULD be set from the `description` field of the `SKILL.md` YAML frontmatter.

Reviewer 4: nit: Why SHOULD vs MUST? Is there a potential that you might have a description that doesn't match the frontmatter that a user could load that would be deceptively?

Author: I could be convinced. I think if we say MUST then we need to decide what to do when they don't match. Is that an error and you can't use the skill? If it is an error, it feels like a bit of an unnecessary footgun. If it's not an error, is the MUST accomplishing anything?

Reviewer 6: I don't we gain anything by enforcing a mimeType, and we add compatible with patterns like zipped skills if we don't. Clients will still validate skills. I think SHOULD serves the purpose. It goes for all resources that they should have the correct mimeType for what they actually are.

### Reading

Skill files are read via the standard `resources/read` method. No skill-specific read semantics are defined.

Internal references within a skill (e.g., `SKILL.md` linking to `references/GUIDE.md`) are relative paths, as in the filesystem form of the Agent Skills specification. A client resolves a relative reference against the skill's root — `references/GUIDE.md` in `skill://acme/billing/refunds/SKILL.md` resolves to `skill://acme/billing/refunds/references/GUIDE.md` — exactly as a filesystem path would resolve. The skill's root is the directory containing `SKILL.md`, not the `skill://` scheme root.

## Implementation Guidelines

The following are recommendations for interoperable implementations. They are not part of the normative specification.

### Hosts: Model-Driven Resource Loading

Hosts SHOULD expose a tool to the model that reads MCP resources by server and URI, enabling the model to load skill content on demand:

```json
{
  "name": "read_resource",
  "description": "Read an MCP resource from a connected server.",
  "inputSchema": {
    "type": "object",
    "properties": {

>       "server": { "type": "string", "description": "Name of the connected MCP server" },

Reviewer 4: AFAIK, it's atypical for the client to expose the server names to the agent. 

The way this is handled for tools, is server names are prefixed only if there is a conflict. Would it be better to follow this standard and only expose resource names, and prefix server name if a conflict occurs?

Author: FWIW, this section is just guidelines and clients are free to expose resource reading in whatever way they find appropriate. The intent was to just guide people to consider that URIs could conflict, so you need some server disambiguation. I could make that clearer. Your suggestion is also fine.

Reviewer 6: @kurtisvg This used to be atypical, but I've noticed that with many clients prefixing server names to tools, the models will infer the name of the server quite often - even if it is not explicitly stated.

As for resource reading - they differ from tools in that that aliasing via name-spacing is awkward, so in many cases a client-side list resource tool will include the server name, as otherwise resolving a resource name conflict is impossible.

So I would agree if this was about tools, where server names are very rarely explicitly stated, but for resources it seems the pattern is to expose server names. At least in the agents I'm most familiar with.

```python
@server.skill("git-workflow")                 # → skill://git-workflow/SKILL.md
def git_workflow():
    return Path("./skills/git-workflow")

@server.skill("acme/billing/refunds")         # → skill://acme/billing/refunds/SKILL.md
def refunds():
    return Path("./skills/refunds")
```

The SDK handles: reading `SKILL.md` frontmatter to populate resource metadata, registering a `skill://<skill-path>/{+path}` resource template, serving file content on `resources/read`, and (where the server's skill set is bounded) responding to scoped `resources/list` calls.

**Client-side** — enumerate and fetch skills:

```python
skills = await client.list_skills()               # wraps resources/list(uri="skill://"), may be empty
content = await client.read_skill_uri(
    "skill://acme/billing/refunds/SKILL.md")      # wraps resources/read, works regardless of enumeration
```

These wrappers are thin — each is a single underlying protocol call with a fixed URI pattern — but they give server authors an ergonomic way to declare skills and give client authors a discoverable entry point.

## Rationale

### Why Resources Instead of a New Primitive?

> The Interest Group's [decision log](decisions.md#2026-02-26-prioritize-skills-as-resources-with-client-helper-tools) records this as settled. Skills are files; Resources exist to expose files. Reusing Resources inherits URI addressability, `resources/read`, `resources/subscribe`, templates, and the existing client tooling for free. A new primitive would duplicate most of this and add [ecosystem complexity the community has explicitly pushed back on](https://github.com/modelcontextprotocol/experimental-ext-skills/issues/14).

Reviewer 4: My 2c is it feels like there's more complexity being added by the use of resources for everything vs and a new primitive. (I also think that 1 open issue is not necessarily "community pushback"). 

E.g. using resources as the distribution mechanism, we lose the ability to have resources that are used for other things being included in a skill. We also have to list every file in skill individually, before the skill is even activated (which is very verbose). There are also the conflicts with how resources are used in existing applications today.

In contrast, a new primitive is clear and unambiguous in it's intent, and could include non-file objects such as tools, resources, and prompts.

Reviewer: > My 2c is it feels like there's more complexity being added by the use of resources for everything vs and a new primitive. (I also think that 1 open issue is not necessarily "community pushback").

I agree that 1 issue is sparse evidence on its own, but the sentiment behind it seems common.

> E.g. using resources as the distribution mechanism, we lose the ability to have resources that are used for other things being included in a skill.

I'm not sure this follows? The `skill://` URI scheme scopes skill content specifically. A server can still expose other resources under any other URI scheme (`file://`, `db://`, etc.), and skills can reference those as dependencies in their frontmatter or content. The skill is a resource and the things it orchestrates like tools, other resources, prompts are still separate and fully available.

> There are also the conflicts with how resources are used in existing applications today.

Could you elaborate on which specific conflicts you're seeing? it seems like structured markdown context fits naturally within modeling other entities like databases, ui views, repos, etc. The `skill://` URI scheme exists to scope skill content away from other resource usage.

> In contrast, a new primitive is clear and unambiguous in it's intent, and could include non-file objects such as tools, resources, and prompts.

A skill referencing tools, resources, and prompts as dependencies seems like it would work the same way regardless of whether the skill itself is delivered as a resource or a new primitive. The SKILL.md frontmatter declares what the skill needs and the host mediates availability. The transport mechanism for the skill content doesn't seem like it would change this.

Also noting a couple of the [MCP design principles](https://modelcontextprotocol.io/community/design-principles) that seem especially relevant here:

- **Composability over specificity**: "MCP provides foundational primitives: resources, tools, prompts, and tasks. We don't add protocol features for use cases that can be constructed from these existing building blocks." Skills are structured context for language models, which seems like exactly what resources are designed to deliver.

- **Standardization over innovation**: "MCP standardizes patterns that have already proven valuable. We look for conventions that work across multiple implementations and codify them, rather than inventing new paradigms." Multiple implementations have independently referenced `skill://` resources, and it seems like skills are commonly used because they're just files (a.k.a. simple resources) and not an abstract concept.

I think the bar for introducing a new primitive should be demonstrating something that resources genuinely can't do based on evidence in real implementations.  Part of this proposal allows for experimenting with resources to be more accessible in general, for example by suggesting common client tools to interact with them.  It also allows clients to follow progressive disclosure patterns and prioritize resources using existing annotations (which follows the other unique properties of skills). If there are specific capabilities we think require a new primitive, would love to dig into those concretely.

# SEP-0000: Skills Extension

- **Status**: Draft (pre-submission)
- **Type**: Extensions Track
- **Created**: 2026-03-17
- **Author(s)**: Skills Over MCP Interest Group
- **Sponsor**: _(seeking)_
- **Extension Identifier**: `io.modelcontextprotocol/skills`
- **PR**: _(to be assigned on submission)_

> This document is a pre-submission draft maintained by the [Skills Over MCP Interest Group](https://github.com/modelcontextprotocol/experimental-ext-skills). It has not yet been submitted to the main MCP repository. Discussion welcome via [GitHub Issues](https://github.com/modelcontextprotocol/experimental-ext-skills/issues) or [Discord #skills-over-mcp-ig](https://discord.com/channels/1358869848138059966/1464745826629976084).

> ## Abstract

Reviewer 6: I think we should mention that the skill format has design that facilitates progressive disclosure, whether this is implemented and if so, how, is up to the client host.

I think we should use client host over agent to match general MCP verbiage. I could be mistaken, but I don't think we use agent elsewhere to describe client host? (It might actually be a good idea, though - as client and client host are so often conflated).

Reviewer: Looks like this was incorporated into the draft but just didn't respond to the comment - resolving this one.

> This SEP defines a convention for serving [Agent Skills](https://agentskills.io/) over MCP using the existing Resources primitive. A _skill_ is a directory of files (minimally a `SKILL.md`) that provides structured workflow instructions to an agent. This extension specifies that each file in a skill directory is exposed as an MCP resource under the `skill://` URI scheme. Skills are addressed by URI and may be read directly; enumeration via `resources/list` and discovery via resource templates are supported but not required, accommodating servers whose skill catalogs are large, generated, or otherwise unenumerable. The skill format itself — directory structure, YAML frontmatter, naming rules — is delegated entirely to the [Agent Skills specification](https://agentskills.io/specification); this SEP defines only the transport binding.

Reviewer 6: Suggest changing directory to folder to match agentskills.io verbiage.

```
skill://<skill-path>/<file-path>
```

where:

- `<skill-path>` is a `/`-separated path of one or more segments locating the skill directory within the server's skill namespace. It MAY be a single segment (`git-workflow`) or nested to arbitrary depth (`acme/billing/refunds`).
- `<file-path>` is the file's path relative to the skill directory root, using `/` as the separator.

The resource for the skill's required `SKILL.md` is therefore always addressable as `skill://<skill-path>/SKILL.md`, and the skill's root directory is the URI obtained by stripping the trailing `SKILL.md`.

The final segment of `<skill-path>` MUST equal the skill's `name` as declared in its `SKILL.md` frontmatter. Preceding segments, if any, are a server-chosen organizational prefix — servers MAY organize skills hierarchically by domain, team, version, or any other axis. In `skill://acme/billing/refunds/SKILL.md`, the prefix is `acme/billing` and the skill's `name` is `refunds`; in `skill://git-workflow/SKILL.md` there is no prefix and the `name` is `git-workflow`. This means the skill name is always recoverable from the URI alone, without reading frontmatter.

Further constraints:

> - A `SKILL.md` MUST NOT appear in an ancestor directory of another `SKILL.md` under the same `skill://` root. The skill directory is the boundary; skills do not nest inside other skills.

Reviewer 7: ```suggestion
- A `SKILL.md` MUST NOT appear in any descendant directory of a skill under the same `skill://` root. The skill directory is the boundary; skills do not nest inside other skills.
```

Reviewer 7: I took the meaning, but still, this line felt less concise than it should be.

| `acme/billing/refunds` | `templates/email.md` | `skill://acme/billing/refunds/templates/email.md` |

#### Resource Metadata

For each `skill://<skill-path>/SKILL.md` resource:

- `mimeType` SHOULD be `text/markdown`.
- `name` SHOULD be set from the `name` field of the `SKILL.md` YAML frontmatter. By the path constraint above, this will equal the final segment of `<skill-path>`.
- `description` SHOULD be set from the `description` field of the `SKILL.md` YAML frontmatter.

Servers MAY expose additional frontmatter fields via the resource's `_meta` object. Other files in the skill use the `mimeType` appropriate to their content.

### Discovery

A server is not required to make its skills enumerable. A `skill://` URI is directly readable via `resources/read` whether or not it appears in any list response, and hosts MUST support loading a skill given only its URI (see [Hosts: Model-Driven Resource Loading](#hosts-model-driven-resource-loading)). This is the baseline: if a model has the URI — from server instructions, from another skill, from the user — it can read the skill.

On top of that baseline, three discovery mechanisms are defined. A server MAY support any combination.

> #### Enumeration via `resources/list`

Reviewer 8: Should we have a initialize as well which is called on server initialize where the client would know what this Skills MCP server is about and what it contains? 
Example, I might have a 3 Skills MCP servers included and which server brings which skills would be important without loading all the list of skills

Reviewer 6: I don't think we should tie this explicitly to connection initialization (I assume that is what you reference). Tools also leave it up to the client when to do this. Having primitives enumerated on server initialized isn't done for any other primitives.

This would also be in conflict with stateless MCP where there are no connections. Enumeration should be a host-design decision. (in practice it often happens after initializing the connection, but it shouldn't have to)

#### Capability Declaration

Per [SEP-2133] extension negotiation, servers declare support for this extension in their `initialize` response:

```json
{
  "capabilities": {
    "extensions": {
      "io.modelcontextprotocol/skills": {}
    }
  }
}
```

No extension-specific settings are currently defined; an empty object indicates support.

### Reading

Skill files are read via the standard `resources/read` method. No skill-specific read semantics are defined.

Internal references within a skill (e.g., `SKILL.md` linking to `references/GUIDE.md`) are relative paths, as in the filesystem form of the Agent Skills specification. A client resolves a relative reference against the skill's root — `references/GUIDE.md` in `skill://acme/billing/refunds/SKILL.md` resolves to `skill://acme/billing/refunds/references/GUIDE.md` — exactly as a filesystem path would resolve. The skill's root is the directory containing `SKILL.md`, not the `skill://` scheme root.

## Implementation Guidelines

The following are recommendations for interoperable implementations. They are not part of the normative specification.

### Hosts: Model-Driven Resource Loading

> Hosts SHOULD expose a tool to the model that reads MCP resources by server and URI, enabling the model to load skill content on demand:

Reviewer 6: I think we need to mention frontmatter loading if we begin to give guidance on implementation.

Something like

Hosts SHOULD load the frontmatter of all available and enabled tools into model context in an implementation-specific manner, so that the model has the server and URI information to be able call the exposed tool when a skill is relevant.

Hosts SHOULD make expose the available tools to the user so they can be inspected and enabled/disabled in a similar manner to tools.

Reviewer: (This was added)

> ```json

Reviewer 8: Should we include a standard response format to share what dependencies and scopes that might be required to run this skill?

```
{
  "jsonrpc": "2.0",
  "id": 2,
  "result": {
    "contents": [
      {
        "uri": "file:///project/src/main.py",
        "mimeType": "text/x-python",
        "dependencies": [
          {
            "name": "requests",
            "version": ">=2.0.0",
            "description": "Used to make HTTP requests to external APIs such as GitHub"
          }
        ],
        "scopes": [
          {
            "name": "github:repo:read",
            "description": "Allows reading repository metadata and contents from GitHub"
          }
        ],
        "text": "# Example: Fetch repository details from GitHub API\n\nimport requests\n\nrepo = \"octocat/Hello-World\"\nurl = f\"https://api.github.com/repos/{repo}\"\n\nresponse = requests.get(url)\nif response.status_code == 200:\n    data = response.json()\n    print(f\"Repo: {data['full_name']}\")\n    print(f\"Stars: {data['stargazers_count']}\")\nelse:\n    print(\"Failed to fetch repository data\")\n"
      }
    ]
  }
}
```

It might add new optional primitives to the MCP spec for resources. But it might be good to add them?

Reviewer: Hi @sunishsheth2009 , sorry you didn't get a direct response on this one (although I think we might have chatted about this in the office hours when you were there)?   Since protocol changes are out of scope for this particular SEP, I think this idea/question should have its own discussion outside of this PR.

{
  "name": "read_resource",
  "description": "Read an MCP resource from a connected server.",
  "inputSchema": {
    "type": "object",
    "properties": {
      "server": { "type": "string", "description": "Name of the connected MCP server" },
      "uri":    { "type": "string", "description": "The resource URI, e.g. skill://git-workflow/SKILL.md" }
    },
    "required": ["server", "uri"]
  }
}
```

Including the server name disambiguates identical `skill://` URIs served by different connected servers. This tool is general-purpose — it reads any MCP resource — and benefits resource use cases beyond skills.

> A typical flow: the host calls `resources/list(uri="skill://")` on each connected server at initialization and surfaces any returned skill entries in the model's context. The model calls `read_resource` with a concrete URI — one returned by enumeration, one handed to it by the user (who may have found it via a `skill://` resource template in the host UI), or one obtained out-of-band — when a skill is relevant to the task.

Reviewer 8: > the host calls `resources/list(uri="skill://")` on each connected server at initialization and surfaces any returned skill entries in the model's context.

Isn't this causing a lot of bloat for the LLM that it would read all the skills even though its not required? Like for a question about git-workflows the LLM would only want to list git-workflows and not skills for acme workflow

Reviewer 6: The results of the list call aren't put into context as-is, the frontmatter is. Most hosts would make it possible to enable/disable skills - similar to tools. And the frontmatter is required for a model to know if a skill is relevant to load, so if it is not in the frontmatter it cannot be accessed at all - regardless of relevance, unless it done in a non-skill like way (ie direct resource reads without progressive disclosure).

Reviewer 8: Got it. I am curious to know how would use this in practice when we want to wrap a stateless agent with MCP servers. Example: I have build a marketing agent which has access to Facebook skills and tools + Youtube Skills and tools. When a question comes in about facebook marketing, how would the agent know to only get facebook skills and tools into the context and not anything else without know which MCP server is for what without reading the tools description. The way that it works right now is that agents call the [initialization](https://modelcontextprotocol.io/specification/2025-06-18/basic/lifecycle#initialization) here to know the description of the MCP server to know what this server does. 
But I am curious if this can be solved differently. :) thank you for your insights

## Backward Compatibility

This extension introduces no new protocol methods, message types, or schema changes beyond those already proposed in [SEP-2093]. A server that does not implement this extension simply exposes no `skill://` resources; existing clients are unaffected. A client that does not implement this extension sees `skill://` resources as ordinary resources, which they are.

Existing implementations using other `skill://` URI structures (NimbleBrain's `skill://server/skill`, skilljack's implicit-`SKILL.md` `skill://name`) will need to adjust their URI paths to conform. These are small, mechanical changes.

## Reference Implementation

Will be provided prior to reaching Final status.

## Security Implications

Skill content is instructional text delivered to a model, which makes it a prompt-injection surface. The Interest Group's position, recorded in [open-questions.md §10](open-questions.md#10-how-should-skills-handle-security-and-trust-boundaries), is:

> - **Trust inherits from the server.** A user who connects a server has already extended their trust boundary to it; a malicious server can cause more harm via tools than via a skill document. Skills do not introduce a new trust tier.

Reviewer 7: ```suggestion
- **Trust inherits from the server.** A user who connects a server has already extended their trust boundary to it; a malicious server can cause as much harm via tools as via a skill document. Skills as a class, whether served by MCP or any other agent affordance, have an inherent prompt injection risk. No additional risk from serving them as MCP resources has been identified.
```

Reviewer 7: The claim of tools posing more of a risk than skills seems hard to substantiate; skills literally tell the model how to behave. 

IMO, it's best to not downplay the risk but to make it clear that Skills over MCP isn't adding any new risk to what already exists with skills.
